### PR TITLE
C51-223: "Add New" case detail dropdown

### PR DESCRIPTION
**Pull request by reneolivo**
This PR adds a "Add New" dropdown button to the case details menu. This dropdown has multiple actions for creating new activities and timelines, sending emails, creating PDFs, etc.

## How it looks

record

### Dropdown directive

A new dropdown directive had to be introduced because Bootstrap and UI Bootstrap don't support nested dropdowns any longer (for more information please read the following issue: https://github.com/twbs/bootstrap/issues/16387#issuecomment-97153831).

This dropdown also allows elements inside dropdown menus to be interactive. Previously the dropdown menu would close as soon as anything was clicked.

A sample template to create a new dropdown could be as the following one:

```html
<span civicase-dropdown>
  <button type="button" civicase-dropdown-toggle>
    Open parent dropdown
  </button>
  <ul class="dropdown-menu">
    <li><a href>Parent option 1</a></li>
    <li><a href>Parent option 2</a></li>
    <li><a href>Parent option 3</a></li>
    <li civicase-dropdown civicase-dropdown-trigger="hover"> <!-- allows the dropdown to show/hide on hover -->
      <a civicase-dropdown-toggle href>Open child dropdown</a>
      <ul class="dropdown-menu">
        <li><a href>Child option 1</a></li>
        <li><a href>Child option 2</a></li>
        <li><a href>Child option 3</a></li>
      </ul>
    </li>
  </ul>
</span>
```

The implementation of the directive is the following one:

```js
function civicaseDropdownLink (scope, $element, attrs) {
  var $dropdownMenu = $element.find('.dropdown-menu:first');
  var $toggleElement = $element.find('[civicase-dropdown-toggle]:first');
  var isOpen = false;

  (function () {
    initEventListeners();
    removeEventListenersOnScopeDestroy();
  })();

  function closeDropdownOnClickOutside (event) {
    var targetIsButtonOrLink = $(event.target).is('button, a');
    var targetIsInsideDropdown = $dropdownMenu[0].contains(event.target);
    var targetIsInsideContainer = $element[0].contains(event.target);
    var targetIsADropdownToggleButton = $(event.target).is('[civicase-dropdown-toggle]');

    if (!targetIsInsideContainer || (targetIsInsideDropdown && targetIsButtonOrLink && !targetIsADropdownToggleButton)) {
      isOpen = false;
      $dropdownMenu.hide();
    }
  }

  function closeDropdownOnEscapeKeyPressed (event) {
    if (isOpen && event.key === 'Escape') {
      isOpen = false;

      event.stopImmediatePropagation();
      $dropdownMenu.hide();
    }
  }

  function hideDropdown () {
    isOpen = false;

    $dropdownMenu.hide();
  }

  function initEventListeners () {
    $toggleElement.on('click', toggleDropdown);
    $('body').on('keydown', closeDropdownOnEscapeKeyPressed);
    $(document).on('click', closeDropdownOnClickOutside);

    if (attrs.civicaseDropdownTrigger === 'hover') {
      $element.on('mouseover', showDropdown);
      $element.on('mouseout', hideDropdown);
    }
  }

  function showDropdown () {
    isOpen = true;

    $dropdownMenu.show();
  }

  function removeEventListenersOnScopeDestroy () {
    scope.$on('$destroy', function () {
      $toggleElement.unbind('click', toggleDropdown);
      $('body').unbind('keydown', closeDropdownOnEscapeKeyPressed);
      $(document).unbind('click', closeDropdownOnClickOutside);
      $element.unbind('mouseover', showDropdown);
      $element.unbind('mouseout', hideDropdown);
    });
  }

  function toggleDropdown () {
    isOpen ? hideDropdown() : showDropdown();
  }
}
```

### Case details dropdown

The actual implementation of the dropdown uses the new dropdown directive plus existing code for the activities and timeline creation and filtering:

```html
<ul civicase-case-tab-affix class="civicase__case-body_tab nav nav-pills nav-pills-horizontal nav-pills-horizontal-primary">
  <!-- ... -->
  <li class="pull-right" civicase-dropdown>
    <div class="civicase__case-details__add-new-dropdown">
      <button type="button" class="btn btn-primary" civicase-dropdown-toggle>
        <i class="fa fa-plus-circle"></i>Add new<i class="fa fa-caret-down"></i>
      </button>
      <ul class="dropdown-menu">
        <li>
          <a
            class="crm-popup"
            ng-href="{{ 'civicrm/activity/email/add' | civicaseCrmUrl:{ action: 'add', caseid: item.id, reset: 1, cid: role.contact_id } }}">
            <i class="material-icons">email</i>
            {{ts('Create email')}}
          </a>
        </li>
        <li>
          <a
            class="crm-popup"
            ng-href="{{ 'civicrm/activity/pdf/add' | civicaseCrmUrl:{ action: 'add', caseid: item.id, reset: 1, cid: role.contact_id, context: 'standalone' } }}">
            <i class="material-icons">picture_as_pdf</i>
            {{ts('Create PDF letter')}}
          </a>
        </li>
        <li class="divider"></li>
        <li class="civicase__activity-dropdown" civicase-dropdown civicase-dropdown-trigger="hover">
          <a href civicase-dropdown-toggle>
            {{ts('Add activity')}}
            <i class="fa fa-caret-right pull-right"></i>
          </a>
          <ul class="dropdown-menu civicase__dropdown-menu--filters">
            <li class="form-group has-feedback">
              <input class="form-control" type="text" ng-model="activityTypeSearchText" />
              <i class="fa fa-search form-control-feedback"></i>
            </li>
            <li class="divider"></li>
            <li ng-repeat="actType in availableActivityTypes | filter:{label: activityTypeSearchText}">
              <a crm-popup-form-success="pushCaseData($data.civicase_reload[0])" class="crm-popup" ng-href="{{ newActivityUrl(actType) }}">
                <i class="fa fa-fw {{actType.icon || 'fa-circle'}}"></i>{{actType.label}}
              </a>
            </li>
          </ul>
        </li>
        <li class="civicase__timeline-dropdown" civicase-dropdown civicase-dropdown-trigger="hover">
          <a href civicase-dropdown-toggle>
            {{ts('Add timeline')}}
            <i class="fa fa-caret-right pull-right"></i>
          </a>
          <ul class="dropdown-menu civicase__dropdown-menu--filters">
            <li class="form-group has-feedback">
              <input class="form-control" type="text" ng-model="activitySetSearchText" />
              <i class="fa fa-search form-control-feedback"></i>
            </li>
            <li class="divider"></li>
            <li ng-repeat="timeline in item.definition.activitySets | filter:{label: activitySetSearchText}">
              <a href crm-confirm="{message: ts('Add all activities from %1 to this case?', {1: timeline.label})}" on-yes="addTimeline(timeline.name)">
                {{timeline.label}}
              </a>
            </li>
          </ul>
        </li>
      </ul>
    </div>
  </li>
</ul>
```

The variables and methods used in this template were already available to the case details controller and no new functions had to be defined.

### CODE REVIEW

**Comment by deb1990** [comment]
@reneolivo when I add a timeline, I see no new activities in the activity tab. I think this functionality is not working.

File: ang/civicase/CaseDetails--tabs.html [hunk 1]
```diff
@@ -5,6 +5,54 @@
       <span class="sr-only" ng-if="activeTab === tab.name">{{ ts('Current') }}</span>
     </a>
   </li>
+  <li class="pull-right" civicase-dropdown>
+    <div class="civicase__case-details__add-new-dropdown">
+      <button type="button" class="btn btn-primary" civicase-dropdown-toggle>
+        <i class="fa fa-plus-circle"></i>Add new<i class="fa fa-caret-down"></i>
+      </button>
+      <ul class="dropdown-menu">
+        <li><a class="crm-popup" ng-href="{{ 'civicrm/activity/email/add' | civicaseCrmUrl:{ action: 'add', caseid: item.id, reset: 1, cid: role.contact_id } }}"><i class="material-icons">email</i> {{ts('Create email')}}</a></li>
```
Comment (deb1990) [inline ang/civicase/CaseDetails--tabs.html]: please put this in a new line, to make things more readable
```
<li>
  <a
      class=""
      ng-href="">
  </a>
</li>
```

File: ang/civicase/CaseDetails--tabs.html [hunk 2]
```diff
@@ -5,6 +5,54 @@
...
+        <li><a class="crm-popup" ng-href="{{ 'civicrm/activity/pdf/add' | civicaseCrmUrl:{ action: 'add', caseid: item.id, reset: 1, cid: role.contact_id, context: 'standalone' } }}"><i class="material-icons">picture_as_pdf</i> {{ts('Create PDF letter')}}</a></li>
```
Comment (deb1990) [inline ang/civicase/CaseDetails--tabs.html]: Same as above.

File: ang/civicase/Dropdown.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  var module = angular.module('civicase');
+
+  /**
+   * This dropdown menu directive allows for nested dropdowns to work properly
+   * and also allows interacting with elements inside of the dropdown without closing
+   * it.
+   */
+  module.directive('civicaseDropdown', function () {
+    return {
+      link: function (scope, $element) {
```
Comment (deb1990) [inline ang/civicase/Dropdown.js]: Link should be moved to a separate function.

File: scss/base/_variables.scss [hunk 4]
```diff
@@ -48,6 +48,7 @@ $activities-calendar-active-day-background: #cde1ed;
 $contact-avatar-background: #99c6e5;
 $case-card-hover: #d9edf7;
 $calendar-due-text-color: #a9a9b7;
+$icon-highlight-color: #0071bd;
```
Comment (deb1990) [inline scss/base/_variables.scss]: this is same as `$brand-primary`, we dont need a new variable.

File: ang/civicase/CaseDetails--tabs.html [hunk 5]
```diff
@@ -5,6 +5,54 @@
...
+        <li class="divider"></li>
+        <li class="civicase__activity-dropdown" civicase-dropdown>
+          <a href civicase-dropdown-toggle>
+            {{ts('Add activity')}}
+            <i class="fa fa-caret-right pull-right"></i>
+          </a>
+          <ul class="dropdown-menu civicase__dropdown-menu--filters">
+            <li class="form-group has-feedback">
+              <input class="form-control" type="text" ng-model="activityTypeSearchText" />
+              <i class="fa fa-search form-control-feedback"></i>
```
Comment (deb1990) [inline ang/civicase/CaseDetails--tabs.html]: Material icons needs to be used.

File: ang/civicase/CaseDetails--tabs.html [hunk 6]
```diff
@@ -5,6 +5,54 @@
...
+        </li>
+        <li class="civicase__timeline-dropdown" civicase-dropdown>
+          <a href civicase-dropdown-toggle>
+            {{ts('Add timeline')}}
+            <i class="fa fa-caret-right pull-right"></i>
+          </a>
+          <ul class="dropdown-menu civicase__dropdown-menu--filters">
+            <li class="form-group has-feedback">
+              <input class="form-control" type="text" ng-model="activitySetSearchText" />
+              <i class="fa fa-search form-control-feedback"></i>
```
Comment (deb1990) [inline ang/civicase/CaseDetails--tabs.html]: Material icons needs to be used.

File: ang/civicase/Dropdown.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+            $(document).unbind('click', closeDropdownOnClickOutside);
+          });
+        }
+
+        /**
+         * Toggles the dropdown menu visibility
+         */
+        function toggleDropdown () {
+          var action;
+          isOpen = !isOpen;
```
Comment (deb1990) [inline ang/civicase/Dropdown.js]: Please leave a line break.

File: ang/civicase/CaseDetails--tabs.html (hunk 5 above)
Comment (reneolivo) [reply ang/civicase/CaseDetails--tabs.html]: In the original design the search icon is defined using font-awesome. Not sure if all icons must be defined using material icons, if so we'll need to confirm this change with Jon

**Pull request merged**